File: binarizer.py
from PIL import Image, ImageOps
import cv2
import numpy as np
import subprocess

BLACK = (0, 0, 0)
WHITE = (255, 255, 255)

def binarize(orig):
    
    pixels = orig.load()

    weight, height = orig.size

    for x in range(weight):
        for y in range(height):
            pixel = pixels[x, y]
            
            if sum(pixel) / 3 < 115:
            # if min(pixel) < 80:
                pixels[x, y] = BLACK
            else:
                pixels[x, y] = WHITE
    
    return orig

def binarize_mean(orig, offset = 10, window_size = None):
    if window_size is None:
        window_size = min(orig.size) // 5
    
    img = np.array(orig)
    gray = cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)
    
    # clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(window_size,window_size))
    # gray = clahe.apply(gray)
    
    blur = cv2.blur(gray, (window_size, window_size))
    
    binary = np.where(gray > blur - offset, 255, 0).astype(np.uint8)
    # binary = gray - (blur - offset) + 128
    
    binary = cv2.cvtColor(binary, cv2.COLOR_GRAY2RGB)
    binary = Image.fromarray(binary)
    binary = ImageOps.expand(binary, border=5, fill='white')
    
    return binary


def binarize_smart(orig, offset = 10, window_size = None):
    if window_size is None:
        window_size = min(orig.size) // 5
    
    img = np.array(orig)
    gray = cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)

    clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(window_size,window_size))
    gray = clahe.apply(gray)

    # binary = cv2.adaptiveThreshold(
    #     gray, 
    #     255,
    #     cv2.ADAPTIVE_THRESH_GAUSSIAN_C,
    #     cv2.THRESH_BINARY,
    #     blockSize=35,
    #     C=-10
    # )
    
    binary = gray
    
    binary = cv2.cvtColor(binary, cv2.COLOR_GRAY2RGB)
    binary = Image.fromarray(binary)
    binary = ImageOps.expand(binary, border=5, fill='white')
    
    return binary


def binarize_sauvola(image, window=None, k=0.3, R=128):
    if window is None:
        window = min(image.size) // 5
    
    gray = cv2.cvtColor(np.array(image), cv2.COLOR_RGB2GRAY)

    mean = cv2.boxFilter(gray, cv2.CV_32F, (window, window))
    sqmean = cv2.boxFilter(gray * gray, cv2.CV_32F, (window, window))
    variance = sqmean - mean * mean
    std = np.sqrt(np.maximum(variance, 0))

    thresh = mean * (1 + k * ((std / R) - 1))
    binary = (gray > thresh).astype(np.uint8) * 255

    return Image.fromarray(binary)


def binarize_zxing_cpp(img):
    img.save("zxing-cpp_finder/input.png")
    subprocess.run('.\\"zxing-cpp_finder\\zxing_binarize.exe" "zxing-cpp_finder/input.png"', shell=True)
    output = Image.open("zxing-cpp_finder/output.png")
    return output.convert("RGB")


if __name__ == "__main__":
    
    # orig = Image.open("orig.png")
    orig = Image.open("corrected.png")
    # orig = Image.open("data/v2/H/small/img0039_v2_H_small_d1.png")
    # orig = Image.open("data/v2/H/big/img0017_v2_H_big_d0.png")
    orig = Image.open("C:/Users/user/Downloads/sample.png")
    
    orig = binarize_zxing_cpp(orig)
    orig.save("corrected.png")

File: src/cpp/binarizer.cpp
#include "binarizer.hpp"

#include "BitMatrix.h"
#include "HybridBinarizer.h"
#include "ImageView.h"

#include <opencv2/imgproc.hpp>

#include <memory>

namespace qrscanner {

static cv::Mat to_gray_continuous(const cv::Mat& img)
{
    cv::Mat gray;
    if (img.channels() == 1)
        gray = img.clone();
    else if (img.channels() == 3)
        cv::cvtColor(img, gray, cv::COLOR_BGR2GRAY);
    else if (img.channels() == 4)
        cv::cvtColor(img, gray, cv::COLOR_BGRA2GRAY);
    else
        gray = img.clone();

    if (!gray.isContinuous())
        gray = gray.clone();
    return gray;
}

cv::Mat binarize(const cv::Mat& orig)
{
    cv::Mat gray = to_gray_continuous(orig);
    cv::Mat binary;
    cv::threshold(gray, binary, 115, 255, cv::THRESH_BINARY);
    cv::Mat bgr;
    cv::cvtColor(binary, bgr, cv::COLOR_GRAY2BGR);
    return bgr;
}

cv::Mat binarize_mean(const cv::Mat& orig, int offset, int window_size)
{
    cv::Mat gray = to_gray_continuous(orig);
    if (window_size <= 0)
        window_size = std::max(1, std::min(orig.cols, orig.rows) / 5);

    cv::Mat blur;
    cv::blur(gray, blur, cv::Size(window_size, window_size));

    cv::Mat binary(gray.size(), CV_8UC1);
    for (int y = 0; y < gray.rows; ++y) {
        const uchar* g = gray.ptr<uchar>(y);
        const uchar* b = blur.ptr<uchar>(y);
        uchar* out = binary.ptr<uchar>(y);
        for (int x = 0; x < gray.cols; ++x)
            out[x] = g[x] > b[x] - offset ? 255 : 0;
    }

    cv::Mat bgr;
    cv::cvtColor(binary, bgr, cv::COLOR_GRAY2BGR);
    return bgr;
}

cv::Mat binarize_smart(const cv::Mat& orig, int, int window_size)
{
    cv::Mat gray = to_gray_continuous(orig);
    if (window_size <= 0)
        window_size = std::max(1, std::min(orig.cols, orig.rows) / 5);

    cv::Ptr<cv::CLAHE> clahe = cv::createCLAHE(2.0, cv::Size(window_size, window_size));
    cv::Mat enhanced;
    clahe->apply(gray, enhanced);

    cv::Mat bgr;
    cv::cvtColor(enhanced, bgr, cv::COLOR_GRAY2BGR);
    return bgr;
}

cv::Mat binarize_sauvola(const cv::Mat& image, int window, double k, double R)
{
    cv::Mat gray = to_gray_continuous(image);
    if (window <= 0)
        window = std::max(1, std::min(gray.cols, gray.rows) / 5);

    cv::Mat gray32;
    gray.convertTo(gray32, CV_32F);
    cv::Mat mean, sqmean;
    cv::boxFilter(gray32, mean, CV_32F, cv::Size(window, window));
    cv::boxFilter(gray32.mul(gray32), sqmean, CV_32F, cv::Size(window, window));

    cv::Mat binary(gray.size(), CV_8UC1);
    for (int y = 0; y < gray.rows; ++y) {
        const uchar* g = gray.ptr<uchar>(y);
        const float* m = mean.ptr<float>(y);
        const float* sm = sqmean.ptr<float>(y);
        uchar* out = binary.ptr<uchar>(y);
        for (int x = 0; x < gray.cols; ++x) {
            const double var = std::max(0.0, static_cast<double>(sm[x] - m[x] * m[x]));
            const double stddev = std::sqrt(var);
            const double thresh = m[x] * (1.0 + k * ((stddev / R) - 1.0));
            out[x] = g[x] > thresh ? 255 : 0;
        }
    }
    return binary;
}

cv::Mat binarize_zxing_cpp(const cv::Mat& img)
{
    cv::Mat gray = to_gray_continuous(img);
    ZXing::ImageView view(gray.data,
                          static_cast<int>(gray.total()),
                          gray.cols,
                          gray.rows,
                          ZXing::ImageFormat::Lum,
                          gray.cols,
                          1);
    ZXing::HybridBinarizer bin(view);
    auto matrix = bin.getBlackMatrix();

    cv::Mat binary(gray.rows, gray.cols, CV_8UC1);
    for (int y = 0; y < gray.rows; ++y) {
        uchar* row = binary.ptr<uchar>(y);
        for (int x = 0; x < gray.cols; ++x)
            row[x] = matrix->get(x, y) ? 0 : 255;
    }

    cv::Mat bgr;
    cv::cvtColor(binary, bgr, cv::COLOR_GRAY2BGR);
    return bgr;
}

} // namespace qrscanner
